clarity/contracts/arkadiko-oracle-v2-4.clar

```diff
@@ -0,0 +1,252 @@
+      pyth-result
+      (let (
+        (dia-result (get-dia-info token))
+      )
+        (if (is-ok dia-result)
+          dia-result
+          (let (
+            (custom-result (get-custom-info token))
+          )
+            (if (is-ok custom-result)
+              custom-result
+              (get-manual-info token)
+            )
+          )
+        )
+      )
+    )
+  )
+)
+
+;; ---------------------------------------------------------
+;; Price Helpers
+;; ---------------------------------------------------------
+
+(define-public (get-pyth-info (token (string-ascii 12)))
+  (let (
+    (price-info (map-get? prices-pyth { token: token }))
+  )
+    (if (is-some price-info)
+      (let (
+        (price (try! (get-price-pyth (get feed-id (unwrap-panic price-info)))))
+        (decimals (get decimals (unwrap-panic price-info)))
+      )
+        (ok { last-price: price, last-block: burn-block-height, decimals: decimals })
+      )
+      (err ERR-PRICE-NOT-FOUND)
+    )
+  )
+)
+
+(define-public (get-dia-info (token (string-ascii 12)))
+  (let (
+    (price-info (map-get? prices-dia { token: token }))
+  )
+    (if (is-some price-info)
+      (let (
+        (price (unwrap-panic (get-price-dia (get key (unwrap-panic price-info)))))
+        (decimals (get decimals (unwrap-panic price-info)))
+      )
+        (ok { last-price: price, last-block: burn-block-height, decimals: decimals })
+      )
+      (err ERR-PRICE-NOT-FOUND)
+    )
+  )
+)
+
+(define-public (get-custom-info (token (string-ascii 12)))
+  (let (
+    (price-info (map-get? prices-custom { token: token }))
+  )
+    (if (is-some price-info)
+      (let (
+        (price (try! (get-price-custom (get key (unwrap-panic price-info)))))
+        (decimals (get decimals (unwrap-panic price-info)))
+      )
+        (ok { last-price: price, last-block: burn-block-height, decimals: decimals })
+      )
+      (err ERR-PRICE-NOT-FOUND)
+    )
+  )
+)
+
+(define-public (get-manual-info (token (string-ascii 12)))
+  (let (
+    (price-info (map-get? prices-manual { token: token }))
+  )
+    (if (is-some price-info)
+      (let (
+        (price (get price (unwrap-panic price-info)))
+        (decimals (get decimals (unwrap-panic price-info)))
+      )
+        (ok { last-price: price, last-block: burn-block-height, decimals: decimals })
+      )
+      (err ERR-PRICE-NOT-FOUND)
+    )
+  )
+)
+
+
+;; ---------------------------------------------------------
+;; Oracles
+;; ---------------------------------------------------------
+
+(define-public (get-price-pyth (price-feed-id (buff 32)))
+  (let (
+
+    ;; TODO: Update for mainnet
+    ;; (storage 'SP3R4F6C1J3JQWWCVZ3S7FRRYPMYG6ZW6RZK31FXY.pyth-storage-v3)
+    ;; (oracle-data (try! (contract-call? 'SP3R4F6C1J3JQWWCVZ3S7FRRYPMYG6ZW6RZK31FXY.pyth-oracle-v3 get-price price-feed-id storage)))
+
+    (storage .pyth-storage-v3)
+    (oracle-data (try! (contract-call? .pyth-oracle-v3 get-price price-feed-id storage)))
```

Comment:
Q: I thought pyth was a push oracle, meaning we are responsible for updating the prices (which is often done as part of a user action), is this correct? Or are we just going to rely on other protocols keeping the price up to date (e.g. Zest/Granite)?

Reply:
A: Correct. If we would fully implement pyth, it would mean that we push prices when a vault gets updated. As other protocols already update prices on a regular basis, I don't think it's worth the change.